tests/keygen.c

```diff
@@ -0,0 +1,214 @@
+ /*
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include <libp11.h>
+#include <openssl/conf.h>
+#include <openssl/engine.h>
+#include <openssl/pem.h>
+#include <string.h>
+
+static void
+usage(char* argv[])
+{
+	fprintf(stderr,
+		"%s token_label key_label [module]\n",
+		argv[0]);
+}
+
+static void
+display_openssl_errors(int l)
+{
+	const char* file;
+	char buf[120];
+	int e, line;
+
+	if (ERR_peek_error() == 0)
+		return;
+	fprintf(stderr, "At main.c:%d:\n", l);
+
+	while ((e = ERR_get_error_line(&file, &line))) {
+		ERR_error_string(e, buf);
+		fprintf(stderr, "- SSL %s: %s:%d\n", buf, file, line);
+	}
+}
+
+static int sign_verify_test(EVP_PKEY *priv, EVP_PKEY *pub) {
+	EVP_MD_CTX *mdctx = NULL;
+	int retval = 0;
+	char *msg = "libp11";
+	size_t slen;
+	unsigned char *sig = NULL;
+
+	if (!priv || !pub) {
+		fprintf(stderr, "Where are the keys?\n");
+		return -1;
+	}
+	mdctx = EVP_MD_CTX_create();
+	if(!mdctx) {
+		display_openssl_errors(__LINE__);
+		retval = -2;
+		goto err;
+	}
+	if(1 != EVP_DigestSignInit(mdctx, NULL, EVP_sha256(), NULL, priv)) {
+		display_openssl_errors(__LINE__);
+		retval = -3;
+		goto err;
+	}
+	if(1 != EVP_DigestSignUpdate(mdctx, msg, strlen(msg))) {
+		display_openssl_errors(__LINE__);
+		retval = -4;
+		goto err;
+	}
+	if(1 != EVP_DigestSignFinal(mdctx, NULL, &slen)) {
+		display_openssl_errors(__LINE__);
+		retval = -5;
+		goto err;
+	}
+	if(!(sig = OPENSSL_malloc(sizeof(unsigned char) * (slen)))) {
+		display_openssl_errors(__LINE__);
+		retval = -6;
+		goto err;
+	}
+	if(1 != EVP_DigestSignFinal(mdctx, sig, &slen)) {
+		display_openssl_errors(__LINE__);
+		retval = -7;
+		fprintf(stderr, "Sign fail\n");
+		goto err;
+	}
+	printf("Sign success\n");
+
+	if(1 != EVP_DigestVerifyInit(mdctx, NULL, EVP_sha256(), NULL, pub)) {
+		display_openssl_errors(__LINE__);
+		retval = -8;
+		goto err;
+	}
+	if(1 != EVP_DigestVerifyUpdate(mdctx, msg, strlen(msg))) {
+		display_openssl_errors(__LINE__);
+		retval = -9;
+		goto err;
+	}
+	if(1 == EVP_DigestVerifyFinal(mdctx, sig, slen))
+	{
+		printf("Verify success\n");
+		retval = 0;
+	}
+	else
+	{
+		display_openssl_errors(__LINE__);
+		fprintf(stderr, "Verify fail\n");
+		retval = -10;
+		goto err;
```

Comment:
superfluous.

or also add it to the if-branch (not just the else) if you want to avoid stupid failures on extending the test. :)